# Validation design

The monthly vintage analysis in [sql/11_monthly_originations.sql](../sql/11_monthly_originations.sql)
shows that bad rates vary over time. This motivates evaluating the model on later vintages,
but does not tell us how much predictive performance will change. This notebook examines
two questions: how the training and validation populations differ, and how much restricting
training to past loans affects predictions on the validation period.

What it shows:

- The assessed numeric borrower features have PSI below 0.04, yet their combined information
  distinguishes the periods with an adversarial ROC AUC of 0.67. Including the Lending Club
  verdict raises this to 0.96: small individual shifts can coexist with distinguishable periods.
- On the same validation loans, period-mixed training improves ROC AUC by about 0.002, with
  paired intervals supporting this small gain. Detectable population differences therefore
  translate into a modest performance effect in this comparison.

Every analysis uses the union feature set from notebook 21, which combines borrower data with
Lending Club's verdict, and scores it with LightGBM. Because drift and out-of-time cost concern
the model intended for deployment, the notebook follows this combination rather than repeating
the analysis across feature sets. The adversarial check is run both with and without the verdict
to identify where the drift occurs.

In [1]:
import numpy as np
import pandas as pd
from numpy.typing import NDArray

from IPython.display import display

from credit_risk.data import load_loans
from credit_risk.split import out_of_time_split, random_split
from credit_risk.model import (
    build_lgbm,
    LC_VERDICT_NUMERIC, LC_VERDICT_CATEGORICAL,
    UNDERWRITER_NUMERIC, UNDERWRITER_CATEGORICAL,
)
from credit_risk.evaluate import discrimination_metrics, bootstrap_comparison
from credit_risk.drift import population_stability_index, adversarial_validation

TARGET = "target_bad"

NUMERIC = UNDERWRITER_NUMERIC + LC_VERDICT_NUMERIC
CATEGORICAL = UNDERWRITER_CATEGORICAL + LC_VERDICT_CATEGORICAL

df = load_loans()
train_temp, val_temp, _ = out_of_time_split(df)

print(f"train {len(train_temp)}  validation {len(val_temp)}")

train 375212  validation 154703


## What moved

PSI compares one feature at a time, using bins fixed on the training data. Here it covers numeric
features with enough distinct values and excludes missing observations. The thresholds of 0.10
and 0.25 provide practical reference points for interpreting the size of a shift.

Adversarial validation uses the features together to predict whether a loan belongs to training
or validation. It also includes categorical features and missing-value information. Its ROC AUC
measures how distinguishable the periods are, rather than how well the credit model predicts
default.

In [2]:
enough_values = [col for col in NUMERIC if train_temp[col].nunique() >= 20]
skipped = sorted(set(NUMERIC) - set(enough_values))
print("skipped, too few distinct values:", skipped)

psi_by_feature = {}
for col in enough_values:
    expected = train_temp[col].dropna()
    actual = val_temp[col].dropna()
    psi_by_feature[col] = population_stability_index(expected, actual)

psi = pd.Series(psi_by_feature).sort_values(ascending=False)
psi.round(3)

skipped, too few distinct values: ['acc_now_delinq', 'chargeoff_within_12_mths', 'collections_12_mths_ex_med', 'inq_last_6mths', 'pub_rec_bankruptcies']


int_rate                  0.155
dti                       0.033
credit_history_months     0.014
open_acc                  0.014
revol_util                0.013
active_acct_ratio         0.012
pub_rec                   0.010
fico_range_low            0.008
loan_amnt                 0.007
mths_since_last_delinq    0.004
revol_bal                 0.004
delinq_2yrs               0.003
annual_inc                0.003
loan_to_income            0.002
total_acc                 0.001
tax_liens                 0.000
delinq_amnt               0.000
dtype: float64

In [3]:
with_verdict = adversarial_validation(train_temp, val_temp, NUMERIC, CATEGORICAL)
borrower_only = adversarial_validation(
    train_temp, val_temp, UNDERWRITER_NUMERIC, UNDERWRITER_CATEGORICAL
)

print(f"with the LC verdict: AUC {with_verdict['roc_auc']:.3f}")
print(f"borrower data only:  AUC {borrower_only['roc_auc']:.3f}")
print()
print(with_verdict["importances"].head(5))

with the LC verdict: AUC 0.964
borrower data only:  AUC 0.671

num__int_rate          5299
cat__grade              556
num__fico_range_low     371
num__loan_amnt          293
num__dti                270
dtype: int32


All assessed numeric borrower features have PSI below 0.04, but borrower data together separates
the periods with an adversarial ROC AUC of 0.67. These results are compatible: the classifier can
combine small individual shifts and use information outside the PSI check. The result does not
by itself establish that relationships between features have changed.

The periods become much easier to distinguish when the Lending Club verdict is included:
adversarial ROC AUC rises to 0.96, while `int_rate` has the largest PSI at 0.155. This points to
a more visible change in pricing, without establishing its effect on default prediction.

These comparisons summarise eight years of training against seven months of validation.
Similar aggregate distributions can therefore coexist with changes across individual vintages,
as seen in the monthly SQL analysis.

## What it costs

The periods are distinguishable, but the differences that identify them may have limited
relevance to default prediction. We therefore compare predictive performance directly, first
under the original random and temporal splits, then on a shared evaluation set.

In [4]:
# Fits LightGBM and returns its validation metrics and predictions.
def fit_and_score(
    train: pd.DataFrame,
    val: pd.DataFrame,
) -> tuple[dict[str, float], NDArray[np.float64]]:
    cols = NUMERIC + CATEGORICAL

    pipe = build_lgbm(NUMERIC, CATEGORICAL)
    pipe.fit(train[cols], train[TARGET])

    proba = pipe.predict_proba(val[cols])[:, 1]
    y = val[TARGET].values

    metrics = discrimination_metrics(y, proba)
    return metrics, proba


train_rand, val_rand, _ = random_split(df)

temporal_metrics, temporal_proba = fit_and_score(train_temp, val_temp)
random_metrics, _ = fit_and_score(train_rand, val_rand)

pd.DataFrame({"temporal": temporal_metrics, "random": random_metrics}).T.round(4)

,roc_auc,pr_auc,brier
temporal,0.6967,0.2754,0.1200
random,0.6922,0.2610,0.1141


The ROC AUC scores are close, with the random split slightly lower. However, the models are
evaluated on different populations, so this difference cannot be attributed to training
composition alone. Differences in bad rates and borrower mix also affect the comparison.
The next experiment holds the evaluation loans fixed to make the training comparison clearer.

Holding the evaluation set fixed isolates the effect of the split. Both models are scored on the
same validation slice, while only the training composition changes. One model trains on past
loans, and the other trains on a same-size sample drawn from every vintage except the loans being
scored. The second setup reflects the information available to a random split but not to a model
in production.

In [5]:
pool = df.drop(index=val_temp.index)
n_train = len(train_temp)

last_scored = val_temp["issue_month"].max()
example = pool.sample(n=n_train, random_state=0)
from_the_future = (example["issue_month"] > last_scored).mean()
print(f"share of the mixed sample issued after the validation period: {from_the_future:.1%}")

results = {"past only": (temporal_metrics, temporal_proba)}

for seed in [0, 1, 2]:
    mixed = pool.sample(n=n_train, random_state=seed)
    results[f"period mixed (seed {seed})"] = fit_and_score(mixed, val_temp)

pd.DataFrame({name: result[0] for name, result in results.items()}).T.round(4)

share of the mixed sample issued after the validation period: 32.3%


,roc_auc,pr_auc,brier
past only,0.6967,0.2754,0.1200
period mixed (seed 0),0.6988,0.2789,0.1194
period mixed (seed 1),0.6991,0.2794,0.1193
period mixed (seed 2),0.6986,0.2790,0.1194


The three seeds vary the period-mixed training sample while keeping the validation loans fixed.
Their range shows sensitivity to training composition, but does not measure uncertainty from
the loans being scored. The paired comparisons below assess the precision of each mixed-minus-past
performance difference.

### Uncertainty in the comparisons

We use the paired bootstrap from notebook 21 because all four fitted models score the same validation loans. In each sample, loans are drawn with replacement, keeping each loan's outcome and predictions together across models. The fitted models remain fixed.

Each table shows the difference measured on the full validation set and a 95% confidence interval, given by the 2.5th and 97.5th percentiles of 1,000 bootstrap samples with seed 0. Differences are calculated as period mixed minus past only, so positive values favour period-mixed training for ROC AUC and PR AUC, while negative values favour it for Brier score.

The three tables correspond to the training seeds above. The intervals describe sampling uncertainty within this validation period, conditional on each fitted model; they do not cover future-vintage drift or all training variability.

In [6]:
predictions = {name: result[1] for name, result in results.items()}
comparisons = {
    f"seed {seed}: period mixed - past only": (f"period mixed (seed {seed})", "past only")
    for seed in [0, 1, 2]
}

intervals = bootstrap_comparison(
    val_temp[TARGET], predictions, comparisons, n_bootstrap=1000, seed=0,
)
print(f"{len(val_temp):,} validation loans, 1,000 paired bootstrap samples")
for name in comparisons:
    print(name)
    display(intervals.loc[name].round(4))

154,703 validation loans, 1,000 paired bootstrap samples
seed 0: period mixed - past only


,difference,ci_low,ci_high
metric,,,
roc_auc,0.0021,0.0012,0.0029
pr_auc,0.0036,0.0019,0.0052
brier,-0.0007,-0.0008,-0.0006


seed 1: period mixed - past only


,difference,ci_low,ci_high
metric,,,
roc_auc,0.0024,0.0016,0.0033
pr_auc,0.0040,0.0023,0.0057
brier,-0.0007,-0.0008,-0.0006


seed 2: period mixed - past only


,difference,ci_low,ci_high
metric,,,
roc_auc,0.0019,0.0010,0.0027
pr_auc,0.0036,0.0018,0.0055
brier,-0.0007,-0.0008,-0.0006


For seed 0, the ROC AUC gain from period-mixed training is 0.0021, with a 95% interval of
[0.0012, 0.0029]. The gains range from 0.0019 to 0.0024 across the three training seeds, and all
three ROC AUC intervals lie above zero.

The PR AUC and Brier intervals also favour period-mixed training in all three comparisons.
The differences remain small, so the observed cost of restricting training to past loans is
modest within this validation period.

## Conclusions

Bad rates vary across vintages, but this does not require large shifts in every borrower feature.
The assessed numeric features show little movement individually, while their combined information
still distinguishes training from validation with an adversarial ROC AUC of 0.67. Including the
Lending Club verdict raises this to 0.96, making the temporal change more visible.

This separability does not translate into an equally large performance penalty. On the same
validation loans, period-mixed training improves ROC AUC by about 0.002. Paired intervals support
this modest gain across three seeds, alongside higher PR AUC and lower Brier score. The population
has changed in detectable ways, but the observed benefit of training across periods remains small.

Changes in economic conditions can increase default probabilities even for borrowers with similar features. A model may still rank higher-risk loans above lower-risk ones, while underestimating their probability of default—for example, predicting 10% for a group whose observed bad rate is 15%. ROC AUC measures the quality of the ranking, so it can remain similar even when the predicted probabilities become less accurate. Stable ROC AUC therefore does not guarantee stable calibration, which needs to be assessed separately.

The temporal split remains appropriate because deployment requires predicting later loans from earlier information; the small measured
cost does not remove that requirement.